### ISSUE TITLE
Narrowing issue with singleton type

### ISSUE THREAD

**Issue by mtomassoli** [microsoft/pyright#3154]
The following code doesn't typecheck in Pyright 1.1.227, but it probably should:

```python
from typing import Union
from enum import Enum

class Empty(Enum):
    token = 0
_empty = Empty.token

def func(x: Union[int, Empty] = _empty) -> int:
    if x is _empty:
        raise Exception
    return x

class Foo:
    def func(self, x: Union[int, Empty] = _empty) -> int:
        if x is _empty:
            raise Exception
        return x        # Type Error :(
```

Here's the exact error:
```
>pyright --version
pyright 1.1.227

>pyright bug.py       
...
...\bug.py
  ...\bug.py:17:16 - error: Expression of type "int | Empty" cannot be assigned to return type "int"
    Type "int | Empty" cannot be assigned to type "int"
      "Empty" is incompatible with "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 3.361sec

>mypy bug.py
Success: no issues found in 1 source file
```

**Comment by erictraut** [comment]
A type checker can't assume that the type is narrowable in this case. There have been some talks about adding support in the type system for "sentinal values", but those ideas haven't yet crystalized into a formal proposal.

Pyright will perform narrowing if you change it to one of the following:
`if x is Empty.token`
or
`if x == Empty.token`

**Comment by mtomassoli** [comment]
I missed that, but then there's a better solution:
`_empty: Final = Empty.token
`
Unless I'm misusing `Final`. It works, though.

**Comment by erictraut** [comment]
Ah yes, good point. I should have listed that as a third option.

**Comment by mtomassoli** [comment]
One doesn't even need `Final`:
`
_empty: Literal[Empty.token] = Empty.token
`
